File: 02_javascript/hello_xmas.js
/*
        /\
       /  \
       /  \
      /  o \
      /    \
     / u    \
     --------
        ||
        ||
*/

const pad = (length) => " ".repeat(length);
const treePad = (length) => {
  let padding = "";
  for (let i = 0; i < length; i++) {
    if (Math.random() < 0.1) {
      padding += "o";
    } else {
      padding += " ";
    }
  }
  return padding;
};

function drawGroup(n, len) {
  console.log(`${pad(len - n)}/${treePad(n * 2)}\\`);
  console.log(`${pad(len - n - 1)}/${treePad((n + 1) * 2)}\\`);
}

function drawTrunk(spacing) {
  console.log(`${pad(spacing)}||`);
  console.log(`${pad(spacing)}||`);
}

function drawTree() {
  const groups = 4;
  for (let i = 0; i < groups; i++) {
    drawGroup(i, groups);
  }
  console.log("-".repeat(groups * 2 + 2));

  drawTrunk(groups);
}

drawTree();

File: 04_python/hello_xmas.py
#    /\
#   /  \
#   /  \
#  /  o \
#  /    \
# / u    \
# --------
#    ||
#    ||

from random import random, choice

def drawOrnaments(len: int):
    ornaments="🎈🎊🎀🎁🔔🎶"
    line=""
    i=0
    while i < len:
        if random() < 0.2 and i < len - 1:
            line += choice(ornaments)
            i += 2
        else:
            line += " "
            i += 1
    return line

def drawGroup(n: int, len: int):
    print(f"{" " * (len - n)}/{drawOrnaments(n*2)}\\")
    print(f"{" " * (len - n - 1)}/{drawOrnaments((n+1)*2)}\\")

def drawTrunk(spacing: int):
    print(f"{" " * spacing}||")
    print(f"{" " * spacing}||")

def drawTree(groups: int):
    for n in range(groups):
        drawGroup(n, groups)
    print(f"{"-" * (groups * 2 + 2)}")
    drawTrunk(groups)

drawTree(groups=8)
